### DISCUSSION TITLE
Cocotb equivalent to run_until_complete

### DISCUSSION THREAD

**Discussion by timothyscherer** [cocotb/cocotb#3794] [Q&A] [answered]
I am currently trying to write a BFM for an existing Python API that doesn't use async/await. I have it working such that in one mode, I import the original API, then in another mode I import my implementation for cocotb. This part is working well. 

My only problem now is that, in order to match the API, none of the functions should be async. But I still need to call async functions from within.

I am not very familiar with Python async, but I believe [this](https://github.com/python/cpython/blob/main/Lib%2Fasyncio%2Fbase_events.py#L685) does what I think I need.

Also see [this](https://discuss.python.org/t/calling-coroutines-from-sync-code/23027) for reference (and my inspiration for trying to figure this out).

**Comment by marlonjames** [comment]
Take a look at #3615 and #3650.

**Comment by ktbarrett** [comment, marked as answer]
Is there a reason not to make an adapter around the existing Python BFM API to make them all `async`? That's going to be an easier route than the other way.

```python
class MyAPI:
    def do_stuff(): ...
    def check(): ...

class AsyncMyAPI:
    def __init__(self, *args, **kwargs):
        self.my_api = MyAPI(*args, **kwargs)
    async def do_stuff(self):
        return self.my_api.do_stuff()
    async def check(self):
        return self.my_api.check()
```

[I'm sorry that Python functions have color =/](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/)

**Reply by timothyscherer** [reply]
The API I am trying to adapt to cocotb is quite large (and from an outside vendor), and I wanted to maintain the same usage so the examples and existing tests are easy to use. I was hoping to eliminate the step of adding a bunch of `await` to the existing lab tests.

Also, some of the API has no need to be async wrapped. If I create this wrapper, I will still need to redefine every function, including the non-async variants.

My goal with this is to write tests that are portable between simulation and the lab. It is working great so far. Essentially, the test can only call BFM functions (no direct cocotb commands), then the BFM functionality gets switched out based on the test mode (sim/lab). Very similar to [this](https://github.com/cocotb/cocotb/discussions/3615#discussioncomment-8155841) concept, except my original API (`UartInterface`) contains non-async, but my sim model needs to use cocotb async for some of the methods.

It seems like the `cocotb Scheduler` has most of the same functionality as the `asyncio BaseEventLoop`, is there a limitation of cocotb that prevents this behavior from being implemented?

**Reply by marlonjames** [reply]
They key to understand is that cocotb is completely subordinate to the simulator it is interacting with.
The event loop is inside the simulator, and the only time that the cocotb scheduler runs is inside of a callback function.
Those callback functions are only called when something happens in the simulator that cocotb has registered a callback for.
The abstraction of that in Python is that cocotb Tasks await Trigger objects, and resume when the callback associated with the Trigger event occurs.

If you need a way to run a blocking non-async function that will block while the simulator continues to run, you are going to need it in a separate thread that can synchronize to cocotb's Tasks.
cocotb provides `external` and `function` to do that.

**Reply by ktbarrett** [reply]
What you are looking for is `cocotb.function`, which is designed to turn an `async` function into a blocking function, much like `run_in_executor`.

A small example...
```python
@cocotb.function
async def cocotb_uart_driver(data): ...

def lab_uart_driver(data): ...

@cocotb.test
async def test(_):
    lab_uart_driver([1, 2, 3])
    cocotb_uart_driver([1, 2, 3])
```
Or rather than using it as a decorator, you can use it inline.
```python
async def cocotb_uart_driver(data): ...

def lab_uart_driver(data): ...

@cocotb.test
async def test(_):
    lab_uart_driver([1, 2, 3])
    cocotb.function(cocotb_uart_driver)([1, 2, 3])
```

Sorry the names `external` and `function` do a terrible job describing what they are for.

Also, there are some limitations that Marlon pointed out, and I would add that there is no way currently to have the cocotb scheduler operate from different threads at the same time the simulator is running. All threaded calls made in cocotb *must* finish before control returns to the simulator. That may not be an issue if all you are doing is mastering a UART, but being a slave to one simply will not work as external code cannot wake up the cocotb scheduler or do VPI/VHPI/FLI calls.

**Reply by timothyscherer** [reply]
Thanks for the quick responses, I attempted to use this, since it seems to do exactly what I want.

The call to the lab variant works corrently, but when it came to the cocotb variant I ran into the following:
```
     0.00ns INFO     cocotb.regression                  test failed
                                                        Traceback (most recent call last):
                                                          File "test_function.py", line 13, in test
                                                            cocotb_uart_driver([1, 2, 3])
                                                          File ".venv/lib/python3.11/site-packages/cocotb/decorators.py", line 109, in __call__
                                                            return cocotb.scheduler._queue_function(self._coro(*args, **kwargs))
                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                          File ".venv/lib/python3.11/site-packages/cocotb/scheduler.py", line 677, in _queue_function
                                                            raise RuntimeError("queue_function called from unrecognized thread")
                                                        RuntimeError: queue_function called from unrecognized thread
```

So I made an MRE using the example provided, which encountered the same issue:
```py
import cocotb

@cocotb.function
async def cocotb_uart_driver(data):
    print(f"cocotb_uart_driver: {data}")

def lab_uart_driver(data):
    print(f"lab_uart_driver: {data}")

@cocotb.test()
async def test(_):
    lab_uart_driver([1, 2, 3])
    cocotb_uart_driver([1, 2, 3])
```

Am I missing a step here?

EDIT: I also tried using the second syntax style, using inline instead of decorators, with the same result.

**Reply by ktbarrett** [reply]
I'd need to check but I'm guessing that it checks to see of the thread that called it is registered as an external thread. This could be extended to include the main thread, which isn't currently being considered. 

EDIT: This is exactly what's happening. `self._main_thread` should be added to the check. @timothyscherer I'll open an issue against this and we'll try to get this in for 2.0.

EDIT 2: Actually that's a bigger PITA than I anticipated. `cocotb.function` is *clearly* only meant to be used with `cocotb.external` and no other way. Hacking support for the above suggestion on doesn't seem immediately possible. We have to worry about weird scheduling behavior where we end up forcing a scheduling inside of a running Task, much like the old `cocotb.fork` used to, and then all that ensues.

**Reply by timothyscherer** [reply]
I understand, that is a shame. I will proceed with using the async wrapper class. Fortunately, as it turns out I can make the async class inherit from the original, then override the original functions with async definitions, using super() to call the originals. I am honestly kind of surprised that it works, but it avoids the costly (and potentially error inducing) need to wrap the entire API in my new async class.

Thank you @ktbarrett and @marlonjames for the quick support and advice!
